base/compiler/tfuncs.jl

```diff
@@ -282,7 +275,7 @@ function isdefined_tfunc(@nospecialize(arg1), @nospecialize(sym))
                 return Const(true)
             elseif isa(arg1, Const)
                 arg1v = (arg1::Const).val
-                if !ismutable(arg1v) || isdefined(arg1v, idx) || (isa(arg1v, DataType) && is_dt_const_field(idx))
+                if !ismutable(arg1v) || isdefined(arg1v, idx) || isa(arg1v, DataType)
```

Comment:
The `types` field can be undefined and then get set later.

Reply:
Ah. That's one of those where `is_dt_const_field` returned `true`, though.
Could we by any chance just trigger setting it here?

Reply:
It can also get cleared later

Reply:
So you're saying that
```julia
julia> foo() = isdefined(VersionNumber, :types)
foo (generic function with 1 method)

julia> @code_typed foo()
CodeInfo(
1 ─     return true
) => Bool
```
is bad because `foo` won't get recompiled when `VersionNumber.types` gets cleared and `foo()` will then wrongly still return `true`? That is bad, of course, but nothing this PR introduces. It's the same on 1.0, 1.5, and master. I'm not sure how big an undertaking fixing that would be, considering that some code (e.g. `fieldcount`) might rely on this (and `getfield`) being const-inferred.

Reply:
Okay, I managed to construct a test case for the "`types` may get set later" case:
```julia
g37293(::Type{T}) where T = isdefined(T, :types)
struct T37293; x::Val{g37293(T37293)}; end # call g37293 on incomplete type with not-yet-set types
@test g37293(T37293) == isdefined(T37293, :types)
```
This fails on master, but I have an update to this PR in the working that will fix it.

@vtjnash how can I provoke clearing the `types` field?